docs/data/data-grid/server-side-data/tree-data.md

```diff
@@ -2,14 +2,72 @@
 title: React Server-side tree data
 ---
 
-# Data Grid - Server-side tree data [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')🚧
+# Data Grid - Server-side tree data [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')
 
 <p class="description">Tree data lazy-loading with server-side data source.</p>
 
-:::warning
-This feature isn't implemented yet. It's coming.
+To dynamically load tree data from the server, including lazy-loading of children, you must create a data source and pass the `unstable_dataSource` prop to the Data Grid, as detailed in the [overview section](/x/react-data-grid/server-side-data/).
 
-👍 Upvote [issue #3377](https://github.com/mui/mui-x/issues/3377) if you want to see it land faster.
+The data source also requires some additional props to handle tree data, namely `getGroupKey` and `getChildrenCount`.
+If the children count is not available for some reason, but there are some children, `getChildrenCount` should return `-1`.
 
-Don't hesitate to leave a comment on the same issue to influence what gets built. Especially if you already have a use case for this component, or if you are facing a pain point with the [currently proposed workaround](https://mui.com/x/react-data-grid/tree-data/#children-lazy-loading).
+```tsx
+const customDataSource: GridDataSource = {
+  getRows: async (params) => {
+    // Fetch the data from the server
+  },
+  getGroupKey: (row) => {
+    // Return the group key for the row, e.g. `name`
+    return row.name;
+  },
+  getChildrenCount: (row) => {
+    // Return the number of children for the row
+    return row.childrenCount;
+  },
+};
+```
+
+The following tree data example supports filtering, sorting, and pagination on the server.
+It also caches the data by default.
+
+{{"demo": "ServerSideTreeData.js", "bg": "inline"}}
+
+:::info
+The data source demos use a utility function `useMockServer` to simulate the server-side data fetching.
+In a real-world scenario, you would replace this with your own server-side data fetching logic.
+
+Open the info section of the browser console to see the requests being made and the data being fetched in response.
 :::
+
+## Error handling
+
+For each row group expansion, the data source is called to fetch the children. If an error occurs during the fetch, the grid will display an error message in the row group cell. `unstable_onDataSourceError` is also triggered with the error and the fetch params.
```

Comment:
Yes, the one regarding the cache being part of the data source object has already been [discussed](https://github.com/mui/mui-x/pull/12317#discussion_r1628374799). I think the [argument](https://github.com/mui/mui-x/pull/12317#discussion_r1629949630) that @romgrk shared makes sense. Wdyt?

Moving `unstable_onDataSourceError` to the interface seems a nice idea, I am not insisting on not doing it, but I am a bit skeptical about doing so because of the very nature of this method.

If we take a look at other methods in `GridDataSource` like `getRows`, `getGroupKey`, and `getChildrenCount`, they correspond to the actual data source and the data belonging to it. Due to this, in most cases, the users can easily define the data source object outside the component without the need to worry about stabilizing its reference.

```tsx
const dataSource = {
  getRows: () => {},
  ...otherProperties,
}

function Component() {
  return (
    <DataGridPro
      unstable_dataSource={dataSource}
    />
  )
}
```

**Making `onError` part of the data source:** `onError` would mostly be used to show an error message to the users, which would most likely involve setting a component state. So the users can no longer define the data source outside the component, rather they'll have to wrap it with `useMemo` and manage the dependencies and stuff.

```tsx
function Component() {
  const [error, setError] = React.useState('');
  const dataSource = React.useMemo(() => ({
    getRows: () => {},
    onError: (error) => setError(error.message),
    ...otherProperties,
  }), [setError])
  return (
    <DataGridPro
      unstable_dataSource={dataSource}
    />
  )
}
```

That is why I wanted to avoid making this part of the data source object. I'd love to hear your thoughts around it.

Reply:
Re: cache – thanks for the link, keeping it outside of the data source makes sense to me.
Re: `onError` – I see where you're coming from. I'm OK with keeping it as is.

In the future, we might reconsider the stable reference requirement for the data source prop.
Perhaps the better DX would be to change the `key` prop on the Data Grid when you want to switch to a different data source? Thoughts? cc @romgrk

Reply:
> we might reconsider the stable reference requirement for the data source prop.

Should we do the same for columns (and perhaps rows) props if we decide to do that? Since they are also dependent on the stable reference for certain performance implications. 🤔

Reply:
Usually switching `key` sounds like a hack to me, but I'm not closed to the idea if there are good arguments for it.